File: src/year2019/day11/day11.rs
//! [Day 11: Space Police](https://adventofcode.com/2019/day/11)

use aoc::{ocr, Coord, Grid};
use intcode::{Computer, State};
use rustc_hash::FxHashMap;

#[derive(PartialEq)]
enum Color {
    Black,
    White,
}

impl From<&Color> for i64 {
    fn from(value: &Color) -> Self {
        match value {
            Color::Black => 0,
            Color::White => 1,
        }
    }
}

struct Robot {
    brain: Computer,
    panel: FxHashMap<Coord, Color>,
}

impl Robot {
    const TURN_LEFT: i64 = 0;
    const TURN_RIGHT: i64 = 1;

    const MOVES: &[Coord] = &[Coord::UP, Coord::RIGHT, Coord::DOWN, Coord::LEFT];

    fn new(brain: &Computer) -> Self {
        Self {
            brain: brain.clone(),
            panel: FxHashMap::default(),
        }
    }

    fn paint(&mut self, initial_color: Color) {
        self.brain.reset();

        let mut pos = Coord::ZERO;
        let mut direction = 0;

        self.panel.insert(pos, initial_color);

        loop {
            // by default (coords are missing), panel is black
            let color = self.panel.get(&pos).unwrap_or(&Color::Black);

            // provide the current panel color to the computer
            self.brain.push(i64::from(color));

            // first output is the color to paint the panel
            let paint = match self.brain.run() {
                State::Output(value) => {
                    if value == 0 {
                        Color::Black
                    } else {
                        Color::White
                    }
                }
                State::Halted => break,
                State::Input => panic!(),
            };

            // second output is the direction the robot should turn
            let State::Output(turn) = self.brain.run() else {
                panic!()
            };

            // paint the panel
            self.panel.insert(pos, paint);

            // update the robot direction
            direction = match turn {
                Self::TURN_RIGHT => (direction + 1) % 4,
                Self::TURN_LEFT => (direction + 3) % 4,
                _ => panic!(),
            };

            // move the robot
            pos += Self::MOVES[direction];
        }
    }

    fn to_grid(&self) -> Grid<char> {
        let mut ll = Coord::new(i32::MAX, i32::MAX);
        let mut ur = Coord::new(i32::MIN, i32::MIN);
        for pos in self.panel.keys() {
            ll.x = ll.x.min(pos.x);
            ll.y = ll.y.min(pos.y);
            ur.x = ur.x.max(pos.x);
            ur.y = ur.y.max(pos.y);
        }

        let mut grid = Grid::<char>::with_size(ur.x - ll.x + 1, ur.y - ll.y + 1, '.', ' ');
        for (&pos, c) in &self.panel {
            if c == &Color::White {
                grid[pos + ll] = '#';
            }
        }

        grid
    }

    fn drawing(&self) -> String {
        ocr::scan_5x6(&self.to_grid().to_string())
    }
}

struct Puzzle {
    brain: Computer,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        Self {
            brain: Computer::load(data),
        }
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        let mut robot = Robot::new(&self.brain);
        robot.paint(Color::Black);
        robot.panel.len()
    }

    /// Solve part two.
    fn part2(&self) -> String {
        let mut robot = Robot::new(&self.brain);
        robot.paint(Color::White);
        robot.drawing()
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (usize, String) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

File: src/year2019/day11/day11.py
#!/usr/bin/env python3
# [Day 11: Space Police](https://adventofcode.com/2019/day/11)

import atexit
import sys
import time
from collections import defaultdict
from pathlib import Path

filename = sys.argv[1] if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


sys.path.append(Path(__file__).parent.parent.as_posix())
from intcode.Intcode import Computer  # noqa
from ocr.ocr import ocr  # noqa


class Robot:
    TURN_LEFT = 0
    TURN_RIGHT = 1

    COLOR_BLACK = 0
    COLOR_WHITE = 1

    MOVES = [
        (0, 1),  # up
        (1, 0),  # right
        (0, -1),  # down
        (-1, 0),  # left
    ]

    def __init__(self, program) -> None:
        self.brain = Computer()
        self.brain.load(program)

    def show(self):
        inf = int(1e6)
        minx, maxx, miny, maxy = inf, -inf, inf, -inf
        for (x, y), color in self.panel.items():
            if color == Robot.COLOR_WHITE:
                minx = min(minx, x)
                maxx = max(maxx, x)
                miny = min(miny, y)
                maxy = max(maxy, y)

        lines = []
        for y in range(maxy, miny - 1, -1):
            row = ""
            for x in range(minx, maxx + 1):
                c = ".#"[self.panel.get((x, y), 0)]
                row += c
            lines.append(row)
        return "\n".join(lines)

    def paint(self, initial_color):
        self.brain.flush_io()
        self.brain.start(output_mode="yield")

        self.rx, self.ry = 0, 0
        self.direction = 0  # up

        self.panel = defaultdict(lambda: 0)
        self.painted = set()

        self.panel[(0, 0)] = initial_color

        while True:
            # by default (coords are missing), panel is black
            color = self.panel.get((self.rx, self.ry), Robot.COLOR_BLACK)

            # provide the current panel color
            self.brain.input.append(color)

            # first output is the color to paint the panel
            state = self.brain.resume()
            if state != "yield":
                break

            # second output is the direction the robot should turn
            state = self.brain.resume()
            if state != "yield":
                break

            paint = self.brain.output.popleft()
            turn = self.brain.output.popleft()

            # paint the panel
            self.panel[(self.rx, self.ry)] = paint
            self.painted.add((self.rx, self.ry))

            # update the robot direction
            if turn == Robot.TURN_RIGHT:
                self.direction = (self.direction + 1) % 4
            elif turn == Robot.TURN_LEFT:
                self.direction = (self.direction - 1) % 4
            else:
                raise ValueError

            # move the robot
            dx, dy = Robot.MOVES[self.direction]
            self.rx, self.ry = self.rx + dx, self.ry + dy

        assert state == "halted"


robot = Robot(data)

robot.paint(Robot.COLOR_BLACK)
print(len(robot.painted))

robot.paint(Robot.COLOR_WHITE)
print(ocr(robot.show()))
